nginx-module.cmake

```diff
@@ -17,31 +17,46 @@ list(REMOVE_DUPLICATES NGINX_CONF_ARGS)
 
 ## Only for printing
 list(JOIN NGINX_CONF_ARGS " " NGINX_CONF_ARGS_LOG)
-
 message(STATUS "Nginx configuration args: ${NGINX_CONF_ARGS_LOG}")
 
 set(NGINX_MODULE_FILE ${NGINX_SRC_DIR}/objs/ngx_http_datadog_module_modules.c)
 set_source_files_properties(${NGINX_MODULE_FILE} PROPERTIES GENERATED TRUE)
 
-add_custom_command(
-  OUTPUT ${NGINX_MODULE_FILE}
-  COMMAND ./configure ${NGINX_CONF_ARGS}
-  WORKING_DIRECTORY ${NGINX_SRC_DIR}
-  COMMENT "Configuring nginx"
-)
+include(FetchContent)
 
-target_sources(nginx_module
-  PRIVATE
-    ${NGINX_MODULE_FILE}
-)
+set(NGINX_VERSION "" CACHE STRING "The nginx version")
+if(NGINX_VERSION STREQUAL "")
+    message(FATAL_ERROR "NGINX_VERSION not set")
+endif()
+set(NGINX_URL "https://nginx.org/download/nginx-${NGINX_VERSION}.tar.gz")
 
-target_include_directories(nginx_module PUBLIC
-    ${NGINX_SRC_DIR}/src/event
-    ${NGINX_SRC_DIR}/src/http/modules
-    ${NGINX_SRC_DIR}/src/http
-    ${NGINX_SRC_DIR}/src/os/unix
-    ${NGINX_SRC_DIR}/src/core
-    ${NGINX_SRC_DIR}/src/event/modules
-    ${NGINX_SRC_DIR}/objs
-)
+FetchContent_Declare(
```

Comment:
`NGINX_SRC_DIR` is the way. Allow me to copy/paste a previous comment:
> Now, CMake expects the NGINX source directory as input, specified with `-DNGINX_SRC_DIR`, and configures accordingly. Instead of downloading the source code each time, which isn't ideal given that I work with both the stable and mainline NGINX branch, there are a few advantages to this approach:
>
>1. It's quicker to simply rename the folder or adjust the project configuration to use a different directory rather than waiting for CMake to download the source tarball (unless CMake has some caching mechanism).
>2. By avoiding unnecessary downloads every-time consider it should re-configure the project, I'm contributing to saving the planet (unless CMake has some caching mechanism).
>3. I'm not aware of a .tar.gz file following the HEAD of NGINX.

Reply:
@dmehala, we spoke offline and agreed to have both solutions coexist, so your workflow should be unaffected.

Reply:
Why should we have two solution for one problem? As a user that don't know anything about the module, that will be confusing.

Reply:
Having two solutions for this specific problem is not a concern if it enhances the developer experience and lets each developer adjust their workflow to what best suits them.

> As a user that don't know anything about the module, that will be confusing. 

This is only a problem as long as we don't document both methods adequately and ensure at least one of them is required by default.

> It's added complexity for no benefit.

I don't think it's much complexity at all and the benefit is that users and developers won't need to specifically have the nginx source locally and can easily choose the version they prefer with a cmake variable. Bear in mind this is the way most build systems work today in most languages so it's a mischaracterization to say there is no benefit.

In any case, as mentioned, having both **mutually exclusive** options will allow you to continue using your preferred workflow, but to answer some of your points:

> By avoiding unnecessary downloads every-time consider it should re-configure the project, I'm contributing to saving the planet (unless CMake has some caching mechanism).

The download of the source only needs to happen once, further downloads or updates can be inhibited using `FETCHCONTENT_UPDATES_DISCONNECTED`, more info [here](https://cmake.org/cmake/help/latest/module/FetchContent.html#variable:FETCHCONTENT_UPDATES_DISCONNECTED).

> I'm not aware of a .tar.gz file following the HEAD of NGINX.

For HEAD one can still use `NGINX_SRC_DIR`.